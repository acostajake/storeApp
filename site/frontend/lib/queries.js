import gql from 'graphql-tag';

import { perPage } from '../config';

export const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$title: String!
		$desc: String!
		$image: String
		$largeImg: String
		$price: Int!
	) {
		createItem(
			title: $title
			desc: $desc
			image: $image
			largeImg: $largeImg
			price: $price
		) {
			id
		}
	}
`;

export const DELETE_ITEM_MUTATION = gql`
	mutation DELETE_ITEM_MUTATION($id: ID!) {
		deleteItem(id: $id) {
			id
		}
	}
`;

export const UPDATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$id: ID!
		$title: String
		$desc: String
		$price: Int
	) {
		updateItem(id: $id, title: $title, desc: $desc, price: $price) {
			id
			title
			desc
			price
		}
	}
`;

export const GET_ALL_ITEMS = gql`
	query GET_ALL_ITEMS($skip: Int = 0, $first: Int = ${perPage}) {
		items(first: $first, skip: $skip, orderBy: createdAt_DESC) {
			id
			title
			desc
			image
			largeImg
			price
		}
	}
`;

export const GET_ITEM = gql`
	query GET_ITEM($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			desc
			price
		}
	}
`;

export const GET_ITEM_WITH_IMAGE = gql`
	query GET_ITEM_WITH_IMAGE($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			desc
			largeImg
		}
	}
`;

export const PAGINATION_QUERY = gql`
	query PAGINATION_QUERY {
		itemsConnection {
			aggregate {
				count
			}
		}
	}
`;

// Auth
export const RESET_MUTATION = gql`
	mutation RESET_MUTATION(
		$resetToken: String!
		$password: String!
		$confirmPassword: String!
	) {
		resetPassword(
			resetToken: $resetToken
			password: $password
			confirmPassword: $confirmPassword
		) {
			id
			name
			email
		}
	}
`;

export const REQUEST_RESET_MUTATION = gql`
	mutation REQUEST_RESET_MUTATION($email: String!) {
		requestReset(email: $email) {
			message
		}
	}
`;

export const SIGNUP_MUTATION = gql`
	mutation SIGNUP_MUTATION(
		$name: String!
		$email: String!
		$password: String!
	) {
		signup(name: $name, email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGNIN_MUTATION = gql`
	mutation SIGNIN_MUTATION($email: String!, $password: String!) {
		signin(email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGNOUT_MUTATION = gql`
	mutation SIGNOUT_MUTATION {
		signout {
			message
		}
	}
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
	mutation updatePermissions($permissions: [Permission], $userId: ID!) {
		updatePermissions(permissions: $permissions, userId: $userId) {
			id
			permissions
			name
			email
		}
	}
`;

export const CURRENT_USER = gql`
	query {
		me {
			id
			email
			name
			permissions
		}
	}
`;

export const GET_ALL_USERS = gql`
	query {
		users {
			id
			name
			email
			permissions
		}
	}
`;
