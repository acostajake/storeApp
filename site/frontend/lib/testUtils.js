import casual from 'casual';
import {
	CURRENT_USER,
	PAGINATION_QUERY,
	REQUEST_RESET_MUTATION,
} from './queries';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	image: 'dog-small.jpg',
	title: 'dogs are best',
	desc: 'dogs',
	largeImg: 'dog.jpg',
});

const fakeUser = () => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
});

const fakeOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: `${casual.word}.jpg`,
	title: casual.words(),
	price: 4234,
	quantity: 1,
	desc: casual.words(),
});

const fakeOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [fakeOrderItem(), fakeOrderItem()],
	createdAt: '2018-04 - 06T19: 24: 16.000Z',
	user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	item: fakeItem(),
	user: fakeUser(),
	...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key) {
		return this.store[key] || null;
	}

	setItem(key, value) {
		this.store[key] = value.toString();
	}

	removeItem(key) {
		delete this.store[key];
	}
}

const notSignedInMocks = [
	{
		request: { query: CURRENT_USER },
		result: { data: { me: null } },
	},
];

const signedInMocks = [
	{
		request: { query: CURRENT_USER },
		result: { data: { me: fakeUser() } },
	},
];

const signedInWithCartMocks = [
	{
		request: { query: CURRENT_USER },
		result: {
			data: {
				me: {
					...fakeUser(),
					cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
				},
			},
		},
	},
];

const makeMocksFor = (length) => {
	return [
		{
			request: { query: PAGINATION_QUERY },
			result: {
				data: {
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							__typename: 'count',
							count: length,
						},
					},
				},
			},
		},
	];
};

const requestResetMock = [
	{
		request: {
			query: REQUEST_RESET_MUTATION,
			variables: { email: 'test@fakeemail.com' },
		},
		result: {
			data: {
				requestReset: {
					message: 'success',
					__typename: 'Message',
				},
			},
		},
	},
];

export {
	LocalStorageMock,
	fakeItem,
	fakeUser,
	fakeCartItem,
	fakeOrder,
	fakeOrderItem,
	notSignedInMocks,
	requestResetMock,
	signedInMocks,
	signedInWithCartMocks,
	makeMocksFor,
};
