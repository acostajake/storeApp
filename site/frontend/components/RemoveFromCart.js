import React from 'react';
import { Mutation } from 'react-apollo';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { CURRENT_USER, REMOVE_FROM_CART } from '../lib/queries';

const RemoveItemButton = styled.button`
	font-size: 3rem;
	background: none;
	border: 0;
	&:hover {
		color: ${(props) => props.theme.lightblue};
		cursor: pointer;
	}
`;

const RemoveFromCart = (props) => {
	const updateCache = (cache, payload) => {
		const data = cache.readQuery({ query: CURRENT_USER });
		const cartItemId = payload.data.removeFromCart.id;
		data.me.cart = data.me.cart.filter(
			(cartItem) => cartItem.id !== cartItemId
		);
		cache.writeQuery({ query: CURRENT_USER, data });
	};

	return (
		<Mutation
			mutation={REMOVE_FROM_CART}
			variables={{ id: props.id }}
			update={updateCache}
			optimisticResponse={{
				__typename: 'Mutation',
				removeFromCart: {
					__typename: 'CartItem',
					id: props.id,
				},
			}}
		>
			{(removeFromCart, { loading, error }) => (
				<RemoveItemButton
					title='Delete Item'
					disabled={loading}
					onClick={() => {
						removeFromCart().catch((err) => alert(err.message));
					}}
				>
					Remove
				</RemoveItemButton>
			)}
		</Mutation>
	);
};

RemoveFromCart.propsTypes = {
	id: PropTypes.string.isRequired,
};

export default RemoveFromCart;
