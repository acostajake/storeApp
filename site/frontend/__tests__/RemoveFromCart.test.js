import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { ApolloConsumer } from 'react-apollo';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';

import RemoveFromCart from '../components/RemoveFromCart';
import { CURRENT_USER, REMOVE_FROM_CART } from '../lib/queries';
import { fakeUser, fakeCartItem } from '../lib/testUtils';

global.alert = console.log;

const removeFromCartMocks = [
	{
		request: { query: CURRENT_USER },
		result: {
			data: {
				me: {
					...fakeUser(),
					cart: [fakeCartItem({ id: 'abc123' })],
				},
			},
		},
	},
	{
		request: { query: REMOVE_FROM_CART, variables: { id: 'abc123' } },
		result: {
			data: {
				removeFromCart: {
					__typename: 'CartItem',
					id: 'abc123',
				},
			},
		},
	},
];

describe('<RemoveFromCart/>', () => {
	it('renders and matches snapshot', () => {
		const wrapper = mount(
			<MockedProvider>
				<RemoveFromCart />
			</MockedProvider>
		);
		expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
	});
	it('removes an item on click', async () => {
		let apolloClient;
		const wrapper = mount(
			<MockedProvider mocks={removeFromCartMocks}>
				<ApolloConsumer>
					{(client) => {
						apolloClient = client;
						return <RemoveFromCart id='abc123' />;
					}}
				</ApolloConsumer>
			</MockedProvider>
		);
		const res = await apolloClient.query({
			query: CURRENT_USER,
		});
		const {
			data: { me },
		} = res;
		expect(me.cart).toHaveLength(1);
		expect(me.cart[0].item.price).toBe(5000);
		wrapper.find('button').simulate('click');
		await wait();
		wrapper.update();
		const res2 = await apolloClient.query({
			query: CURRENT_USER,
		});
		const {
			data: { me: me2 },
		} = res2;
		expect(me.cart).toHaveLength(0);
	});
});
