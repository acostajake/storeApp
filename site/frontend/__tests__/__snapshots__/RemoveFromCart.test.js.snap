// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveFromCart/> renders and matches snapshot 1`] = `
<button
  className="RemoveFromCart__RemoveItemButton-dy7x4z-0 hrvlQy"
  disabled={false}
  onClick={[Function]}
  title="Delete Item"
>
  Remove
</button>
`;
